src/perform_step/newmark_perform_step.jl

```diff
@@ -0,0 +1,81 @@
+function initialize!(integrator, cache::NewmarkBetaCache)
+    duprev, uprev = integrator.uprev.x
+    integrator.f(cache.fsalfirst, integrator.uprev, integrator.p, integrator.t)
+    integrator.stats.nf += 1
+    integrator.fsalfirst = cache.fsalfirst
+    integrator.fsallast  = cache.fsalfirst
+    # integrator.fsallast = du_alias_or_new(cache.nlsolver, integrator.fsalfirst)
+    return
+end
+
+@muladd function perform_step!(integrator, cache::NewmarkBetaCache, repeat_step = false)
+    @unpack t, dt, f, p = integrator
+    @unpack upred, β, γ, nlsolver = cache
+
+    # Given Mu'' = f(u,u',t) we need to solve the non-linear problem
+    #   Mu(tₙ₊₁)'' - f(ũ(tₙ₊₁) + u(tₙ₊₁)'' β Δtₙ²,ũ(tₙ₊₁)' + u(tₙ₊₁)'' γ Δtₙ,t) = 0
+    # for u(tₙ₊₁)''.
+    # The predictors ũ(tₙ₊₁) are computed as
+    #   ũ(tₙ₊₁)  = u(tₙ) + u(tₙ)' Δtₙ + u(tₙ)'' (0.5 - β) Δtₙ²
+    #   ũ(tₙ₊₁)' = u(tₙ)' + u(tₙ)'' (1.0 - γ) Δtₙ
+    # such that we can compute the solution with the correctors
+    #   u(tₙ₊₁)  = ũ(tₙ₊₁)  + u(tₙ₊₁)'' β Δtₙ²
+    #   u(tₙ₊₁)' = ũ(tₙ₊₁)' + u(tₙ₊₁)'' γ Δtₙ
+
+    mass_matrix = f.mass_matrix
+
+    # Evaluate predictor
+    dduprev = integrator.fsalfirst.x[1]
+    duprev, uprev = integrator.uprev.x
+    upred_full = ArrayPartition(
+        duprev + dt*(1.0 - γ)*dduprev,
+        uprev  + dt*dt*(0.5 - β)*dduprev + dt*duprev
+    )
```

Comment:
this seems like it will allocate a bunch, right?

Reply:
Indeed. As also pointed out below by you during review we might be able to use `nlsolver.tmp` to get rid of this.